messaginginapp/src/test/java/io/customer/messaginginapp/ui/controller/InAppMessageViewControllerTest.kt

```diff
@@ -0,0 +1,490 @@
+        controller.detachEngineWebView()
+
+        controller.engineWebViewDelegate.shouldBeNull()
+        assertNoInteractions(viewDelegate)
+    }
+
+    @Test
+    fun engineSetup_givenLoadMessage_expectEngineConfigured() {
+        val givenStoreState = InAppMessagingState(
+            siteId = String.random,
+            dataCenter = String.random,
+            environment = GistEnvironment.LOCAL
+        )
+        every { inAppMessagingManager.getCurrentState() } returns givenStoreState
+        val givenMessage = createInAppMessage()
+        val engineWebViewDelegate = mockk<EngineWebViewDelegate>(relaxed = true)
+        controller.engineWebViewDelegate = engineWebViewDelegate
+
+        controller.loadMessage(givenMessage)
+
+        controller.currentMessage shouldBeEqualTo givenMessage
+        val expectedConfig = EngineWebConfiguration(
+            siteId = givenStoreState.siteId,
+            dataCenter = givenStoreState.dataCenter,
+            messageId = givenMessage.messageId,
+            instanceId = givenMessage.instanceId,
+            endpoint = givenStoreState.environment.getEngineApiUrl(),
+            properties = givenMessage.properties
+        )
+        assertCalledOnce { engineWebViewDelegate.setup(expectedConfig) }
+    }
+
+    @Test
+    fun engineSetup_givenStopCalled_expectEngineStopsLoading() {
+        val engineWebViewDelegate = mockk<EngineWebViewDelegate>(relaxed = true)
+        controller.engineWebViewDelegate = engineWebViewDelegate
+
+        controller.stopLoading()
+
+        assertCalledOnce { engineWebViewDelegate.stopLoading() }
+    }
+
+    //endregion
+    //region Engine callback tests
+
+    @Test
+    fun engineListener_givenEngineError_expectDispatchMessageFailedEvent() {
+        val givenMessage = createInAppMessage()
+        controller.setMessageAndRouteForTest(message = givenMessage, route = String.random)
+
+        controller.error()
+
+        verifyOrder {
+            inAppMessagingManager.dispatch(
+                InAppMessagingAction.EngineAction.MessageLoadingFailed(givenMessage)
+            )
+        }
+    }
+
+    @Test
+    fun engineListener_givenEngineErrorAndNoCurrentMessage_expectNoEventDispatched() {
+        controller.error()
+
+        assertNoInteractions(inAppMessagingManager)
+    }
+
+    @Test
+    fun engineListener_givenEngineBootstrapped_expectNoEventDispatched() {
+        controller.bootstrapped()
+
+        assertNoInteractions(inAppMessagingManager)
+    }
+
+    @Test
+    fun engineListener_givenSizeChanged_expectViewListenerReceivesCorrectSizeUpdates() {
+        val screenDensity = Double.random(1.0, 3.0)
+        val givenWidthInDp = Double.random(100.0, 500.0)
+        val givenHeightInDp = Double.random(100.0, 500.0)
+        val givenWidthInPx = (givenWidthInDp * screenDensity).roundToInt()
+        val givenHeightInPx = (givenWidthInDp * screenDensity).roundToInt()
```

Comment:
this doesnt look correct? shouldnt it be

```
val givenHeightInPx = (givenHeightInDp * screenDensity).roundToInt()
```

Reply:
Good catch. I've updated the test to fail in case of such typos.